### PR TITLE
feat: add gaea nas [DET-3553]

### PR DESCRIPTION
**Pull request by liamcli**
## Description
Implement recent NAS method proposed in:
https://arxiv.org/abs/2004.07802

[DET-3553]

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

[DET-3553]: https://determinedai.atlassian.net/browse/DET-3553

### CODE REVIEW

File: examples/experimental/trial/gaea_nas/model_def.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        for a in self.model.arch_parameters():
+            a.requires_grad=False
+        for w in self.model.ws_parameters():
+            w.requires_grad=True
+        loss = self.model._loss(x_train, y_train)
+        self.context._backward(loss)
+        self.context._step_optimizer(self.ws_opt)
+
+        # Train arch parameters
+        for a in self.model.arch_parameters():
```
Comment (liamcli) [inline examples/experimental/trial/gaea_nas/model_def.py:123]: ws_opt and arch_opt optimize disjoint sets of parameters of the model.  the optimizer.zero_grad() call at the end of _step_optimizer(self.ws_opt) does not zero out the gradients computed for the arch_parameters.  

without specifying requires_grad for the two sets of parameters, this throws an error for calling backward multiple times.  it was surprising that self.model.zero_grad() did not solve the issue.  i think allowing users to call zero_grad on the model with solve this issue and make it easier to optimize disjoint sets of variables of the same model.

Comment (liamcli) [reply examples/experimental/trial/gaea_nas/model_def.py:123]: @shiyuann

Comment (shiyuann) [reply examples/experimental/trial/gaea_nas/model_def.py:123]: > ws_opt and arch_opt optimize disjoint sets of parameters of the model. the optimizer.zero_grad() call at the end of _step_optimizer(self.ws_opt) does not zero out the gradients computed for the arch_parameters.

I noticed that you set the aggregation_frequency to be 1. With this setting, `step_optimizer` should call `optimizer.zero_grad()` for users and all the parameters associated with that optimizer should clear out their gradients. Wonder if `optimizer.zero_grad()` was actually called.

> Without specifying requires_grad for the two sets of parameters, this throws an error for calling backward multiple times. it was surprising that self.model.zero_grad() did not solve the issue.

I noticed that you were only using parallel training and not AMP.

This is a constraint with horovod because it needs us to tell how many backward passes are done for each optimizer before running `optimizer.step()` on it. It has this constraint because they want the users to write the right code to avoid unnecessary across-GPU communication in distributed training. There are three methods to design the interface to satisfy this constraint.

1. Letting users tell us how many times they want to call `backward` on losses before they call `step_optimizer` each time in advance. This will make the interface ugly and hard to understand.

2. Assume users call `step_optimizer` right after `backward(loss)` each time and assume users will set `requires_grad` for those parameters they want to calculate gradients on. This loses the flexibility to do manual gradient accumulation but helps users write the right code to avoid unnecessary across-GPU communication.

3. Work around horovod to know how many times `backward` is called on losses for each parameter. This makes calling `backward` and `step_optimizer` flexible enough but users might find d-train not as effective as they thought just because they write the inefficient code.

The 2nd method is what we chose intentionally. So rather than exposing `zero_grad` we probably should explain to users more clearly how we want them to write their code.

Comment (shiyuann) [reply examples/experimental/trial/gaea_nas/model_def.py:123]: I am still looking into if we could support local gradient accumulation and avoid doing unnecessary communication. Supporting `zero_grad` generally looks reasonable to me.

Comment (shiyuann) [reply examples/experimental/trial/gaea_nas/model_def.py:123]: Added a ticket https://determinedai.atlassian.net/browse/DET-3513 and add it to #ml-ag meeting.

File: examples/experimental/trial/gaea_nas/const.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    min_learning_rate: 0
+    scheduler_epochs: 50
+    weight_decay: 3e-4
+    arch_learning_rate: 0.1
+    init_channels: 16
+    layers: 8
+    nodes: 4
+
+resources:
+  slots_per_trial: 2
```
Comment (naren-determined) [inline examples/experimental/trial/gaea_nas/const.yaml:24]: Since this is const.yaml, we don't have slots_per_trial to be more than 1 here?

File: examples/experimental/trial/gaea_nas/genotypes.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+
+PC_DARTS_cifar = Genotype(normal=[('sep_conv_3x3', 1), ('skip_connect', 0), ('sep_conv_3x3', 0), ('dil_conv_3x3', 1), ('sep_conv_5x5', 0), ('sep_conv_3x3', 1), ('avg_pool_3x3', 0), ('dil_conv_3x3', 1)], normal_concat=range(2, 6), reduce=[('sep_conv_5x5', 1), ('max_pool_3x3', 0), ('sep_conv_5x5', 1), ('sep_conv_5x5', 2), ('sep_conv_3x3', 0), ('sep_conv_3x3', 3), ('sep_conv_3x3', 1), ('sep_conv_3x3', 2)], reduce_concat=range(2, 6))
+PC_DARTS_image = Genotype(normal=[('skip_connect', 1), ('sep_conv_3x3', 0), ('sep_conv_3x3', 0), ('skip_connect', 1), ('sep_conv_3x3', 1), ('sep_conv_3x3', 3), ('sep_conv_3x3', 1), ('dil_conv_5x5', 4)], normal_concat=range(2, 6), reduce=[('sep_conv_3x3', 0), ('skip_connect', 1), ('dil_conv_5x5', 2), ('max_pool_3x3', 1), ('sep_conv_3x3', 2), ('sep_conv_3x3', 1), ('sep_conv_5x5', 0), ('sep_conv_3x3', 3)], reduce_concat=range(2, 6))
+
+
+PCDARTS = PC_DARTS_cifar
```
Comment (naren-determined) [inline examples/experimental/trial/gaea_nas/genotypes.py:82]: Is this assignment really necessary?

File: examples/experimental/trial/gaea_nas/model_search.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    channels_per_group = num_channels // groups
+    
+    # reshape
+    x = x.view(batchsize, groups, 
+        channels_per_group, height, width)
+
+    x = torch.transpose(x, 1, 2).contiguous()
+
+    # flatten
+    x = x.view(batchsize, -1, height, width)
```
Comment (naren-determined) [inline examples/experimental/trial/gaea_nas/model_search.py:22]: You can just return this statement instead of adding an extra line, but more a style thing

File: examples/experimental/trial/gaea_nas/const.yaml (hunk 2 above)
Comment (liamcli) [reply examples/experimental/trial/gaea_nas/const.yaml:24]: i'm not a fan of differentiating between const and distributed.  const to me means constant hps instead of a search space.

File: examples/experimental/trial/gaea_nas/genotypes.py (hunk 3 above)
Comment (liamcli) [reply examples/experimental/trial/gaea_nas/genotypes.py]: removed
